salt/auth/file.py

```
@@ -0,0 +1,291 @@
+    config['password_field'] = int(config['password_field'])
+
+    return config
+
+
+def _text(username, password, **kwargs):
+    '''
+    The text file function can authenticate plaintext and digest methods
+    that are available in the :py:func:`hashutil.digest <salt.modules.hashutil.digest>`
+    function. (currently md5, sha256, sha512)
+    '''
+    filename = kwargs['filename']
+    hashtype = kwargs['hashtype']
+    field_separator = kwargs['field_separator']
+    username_field = kwargs['username_field']-1
+    password_field = kwargs['password_field']-1
+
+    # add exception handling for out of range fields
+    with io.open(filename, 'r') as pwfile:
+        for line in pwfile.readlines():
+            fields = line.strip().split(field_separator)
+
+            try:
+                this_username = fields[username_field]
+            except IndexError:
+                log.error('salt.auth.file: username field ({0}) does not exist '
+                          'in file {1}'.format(username_field, filename))
+            try:
+                this_password = fields[password_field]
+            except IndexError:
+                log.error('salt.auth.file: password field ({0}) does not exist '
+                          'in file {1}'.format(password_field, filename))
+
+            if this_username == username:
+                if hashtype == 'plaintext':
+                    if this_password == password:
+                        return True
+                else:
+                    # Exceptions for unknown hash types will be raised by hashutil.digest
+                    if this_password == __salt__['hashutil.digest'](password, hashtype):
+                        return True
+
+                # Short circuit if we've already found the user but the password was wrong
+                return False
+    return False
+
+
+def _htpasswd(username, password, **kwargs):
+    from passlib.apache import HtpasswdFile
```

I misstated my concern. In the event that somebody _does_ configure one of these back-ends but doesn't have the required libraries, I think that we should throw a clean error versus just raising an `ImportError`.